omniduct/databases/base.py

```diff
@@ -114,81 +115,61 @@ def execute(self, statement, query=False, parse=True, index_field=None, date_fie
         A DBAPI2 cursor object if `query` is `True`, and `parse` is `False`.
         `None` otherwise.
         '''
+
         self.connect()
+
+        if template:
+            statement = self.render_template(statement, template_namespace, by_name=False)
+
         statements = self.statements_split(statement)
-        statements = [self.statement_cleanup(stmt) if cleanup_statement else stmt for stmt in statements]
+        statements = [self.statement_cleanup(stmt) if cleanup else stmt for stmt in statements]
         assert len(statements) > 0, "No non-empty statements were provided."
-        if render_only:
-            return ';\n'.join(statements)
-        cursor = None
+
         for statement in statements[:-1]:
-            cursor = self.connect()._execute(statement, query=False, cursor=cursor, **kwargs)
-        cursor = self.connect()._execute(statements[-1], query, cursor=cursor, **kwargs)
-
-        if not query or self._cursor_empty(cursor):
-            return None
-        if parse:
-            df = self._cursor_to_dataframe(cursor)
-            cursor.close()
-
-            if date_fields is None:  # if user supplied, use as is
-                date_fields = config.date_fields or []
-                date_fields = [field for field in date_fields if field in df]
-
-            if date_fields:
-                try:
-                    df = pandas.io.sql._parse_date_columns(df, date_fields)
-                except:
-                    logger.error('Unable to parse date columns. Perhaps your version of pandas is outdated.')
-            if index_field is not None:
-                df.set_index(index_field, inplace=True)
-            return df
-        else:
-            return cursor
+            cursor = self.connect()._execute(statement, cursor=cursor, async=False, **kwargs)
+        cursor = self.connect()._execute(statements[-1], cursor=cursor, async=async, **kwargs)
 
-    def query(self, query, **kwargs):
+        return cursor
+
+    @logging_scope("Query", timed=True)
+    @cached_method(
+        id_str=lambda self, kwargs: "{}:\n{}".format(kwargs['format'], self.statement_hash(kwargs['statement'])),
+    )
+    def query(self, statement, format='pandas', format_opts={}, **kwargs):
         '''
         This method  is shorthand for:
-        >>> client.execute(query, query=True, **kwargs)
+        >>> client.execute(statement, **kwargs)
         '''
-        return self.execute(query, query=True, **kwargs)
 
-    def query_dump(self, query, file, format='hive', **kwargs):
-        '''
-        This method executes a query, and streams the results into a file. This is
-        especially useful for large queries, which may not fit entirely in memory.
-        Currently all fields are treated as strings.
+        cursor = self.execute(statement, async=False, **kwargs)
+        formatter = self._get_formatter(format, cursor, **format_opts)
```

Comment:
I opted not to do this so that the intermediate Formatter object had a name. I felt it made it easier to read, rather than just adding `.dump()` to an already long line of code.